{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4222d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seq= \"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\"\n",
    "# new_seq=\"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTCCCCCTCCCCCGACTCGATTGTACGTGAAGCTCCCCCTCCCCCGATCAAGCACACAAAAAACCAACACACAGATCTAATGAAAATAAAGATCTTTTATTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\"\n",
    "# inserted_seq=\"CCCCCTCCCCCGACTCGATTGTACGTGAAGCTCCCCCTCCCCCGATCAAGCACACAAAAAACCAACACACAGATCTAATGAAAATAAAGATCTTTTATT\"\n",
    "\n",
    "# # Open the modified mm10 FASTA file\n",
    "# with open(genome_file, \"r\") as fasta_file:\n",
    "#     # Use SeqIO.parse to read the sequences from the file\n",
    "#     sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "#     # Loop through each sequence in the file\n",
    "#     for seq_record in sequences:\n",
    "#         # Check if this is the sequence you modified\n",
    "#         if seq_record.id == \"2\":\n",
    "#             # Get the length of the inserted sequence\n",
    "#             insertion_length = len(inserted_seq)\n",
    "\n",
    "#             # Open the GTF file\n",
    "#             with open(gtf_file, \"r\") as gtf_file:\n",
    "#                 # Use GFF.parse to read the features from the file\n",
    "#                 features = list(GFF.parse(gtf_file))\n",
    "\n",
    "#                 # Loop through each feature in the file\n",
    "#                 for feature in features:\n",
    "#                     # Check if this feature overlaps with the insertion site\n",
    "#                     if feature.seqid == seq_record.id and feature.end >= insertion_site:\n",
    "#                         # Calculate the new start and end positions\n",
    "#                         new_start = feature.start + insertion_length\n",
    "#                         new_end = feature.end + insertion_length\n",
    "\n",
    "#                         # Update the feature with the modified location\n",
    "#                         feature.location = FeatureLocation(new_start, new_end, feature.strand)\n",
    "\n",
    "#             # Write the modified GTF file\n",
    "#             with open(\"modified_mm10.gtf\", \"w\") as output_file:\n",
    "#                 GFF.write(features, output_file)\n",
    "#         else:\n",
    "#             # If this is not the modified sequence, copy it to the new GTF file unmodified\n",
    "#             with open(\"mm10.gtf\", \"r\") as gtf_file:\n",
    "#                 with open(\"modified_mm10.gtf\", \"a\") as output_file:\n",
    "#                     output_file.write(gtf_file.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "022fbfe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from BCBio import GFF\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b43c359b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "105100639\n"
     ]
    }
   ],
   "source": [
    "genome_file = \"mm10_genome/mm10-2020-A-reference-sources/Mus_musculus.GRCm38.dna.primary_assembly.fa\"\n",
    "gtf_file = \"mm10_genome/mm10-2020-A-reference-sources/gencode.vM23.primary_assembly.annotation.gtf\"\n",
    "output_gtf_file = \"updated_mm10.gtf\"\n",
    "output_genome = \"updated_mm10.fa\"\n",
    "\n",
    "\n",
    "seq= \"TGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\"\n",
    "new_seq=\"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTCCCCCTCCCCCGACTCGATTGTACGTGAAGCTCCCCCTCCCCCGATCAAGCACACAAAAAACCAACACACAGATCTAATGAAAATAAAGATCTTTTATTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\"\n",
    "inserted_seq=\"CCCCCTCCCCCGACTCGATTGTACGTGAAGCTCCCCCTCCCCCGATCAAGCACACAAAAAACCAACACACAGATCTAATGAAAATAAAGATCTTTTATT\"\n",
    "# Open the modified mm10 FASTA file\n",
    "modified_sequences=[]\n",
    "with open(genome_file, \"r\") as fasta_file:\n",
    "    # Use SeqIO.parse to read the sequences from the file\n",
    "    sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "    # Loop through each sequence in the file\n",
    "    for seq_record in sequences:\n",
    "        # Check if this is the sequence you modified\n",
    "        if seq_record.id == \"2\":\n",
    "            # Get the length of the inserted sequence\n",
    "            insertion_length = len(inserted_seq)\n",
    "            match = re.search(seq, str(seq_record.seq))\n",
    "            if match:\n",
    "                # Insert the new subsequence after the pattern\n",
    "                insertion_site = match.start()\n",
    "                print(insertion_site)\n",
    "                seq_record.seq = seq_record.seq[:insertion_site]+inserted_seq+seq_record.seq[insertion_site:]\n",
    "\n",
    "        modified_sequences.append(seq_record)\n",
    "                \n",
    "    #             if seq in str(seq_record.seq):\n",
    "    #             # Create a new sequence object with the modified subsequence\n",
    "    #             old_sequence = seq_record.seq\n",
    "    #             new_sequence = Seq(str(old_sequence).replace(seq, new_seq))\n",
    "\n",
    "    #             # Create a new SeqRecord object with the modified sequence\n",
    "    #             modified_seq_record = seq_record\n",
    "    #             modified_seq_record.seq = new_sequence\n",
    "\n",
    "    #             # Add the modified sequence to the list of modified sequences\n",
    "    #             modified_sequences.append(modified_seq_record)\n",
    "    #             print(seq_record.id)\n",
    "    #         else:\n",
    "    #             # Add the unmodified sequence to the list of modified sequences\n",
    "    #             modified_sequences.append(seq_record)\n",
    "\n",
    "\n",
    "    # Open the GTF file\\\n",
    "    \n",
    "    \n",
    "\n",
    "with open(output_genome, \"w\") as output_file:\n",
    "    SeqIO.write(modified_sequences, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f5e5f8f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = \"mm10_genome/mm10-2020-A-reference-sources/gencode.vM23.primary_assembly.annotation.gtf\"\n",
    "placeholder = \"||\"\n",
    "placeholder = \"||\"\n",
    "\n",
    "gtf_df = pd.read_csv(gtf_file, sep='\\t', header=None, comment='#', names=[\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"])\n",
    "# gtf_df[\"attribute\"] = gtf_df[\"attribute\"].str.replace(r'[\";]', placeholder)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f95fe0c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_716277/708619042.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pos_df['end'] += insertion_length\n",
      "/tmp/ipykernel_716277/708619042.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pos_df['end'] += insertion_length\n",
      "/tmp/ipykernel_716277/708619042.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pos_df['start'] += insertion_length\n"
     ]
    }
   ],
   "source": [
    "chr_name = 'chr2'\n",
    "position = insertion_site\n",
    "\n",
    "chr_df = gtf_df[gtf_df['seqname'] == chr_name]\n",
    "pos_df = chr_df[(chr_df['start'] <= position) & (chr_df['end'] > position)]\n",
    "pos_df['end'] += insertion_length\n",
    "gtf_df.update(pos_df)\n",
    "\n",
    "\n",
    "pos_df = chr_df[(chr_df['start'] > position)]\n",
    "pos_df['end'] += insertion_length\n",
    "pos_df['start'] += insertion_length\n",
    "                \n",
    "gtf_df.update(pos_df)\n",
    "gtf_df[\"start\"] = gtf_df[\"start\"].astype(int)\n",
    "gtf_df[\"end\"] = gtf_df[\"end\"].astype(int)\n",
    "\n",
    "# gtf_df.to_csv(output_gtf_file, sep='\\t', header=None, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b05ad3d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the updated DataFrame to a new GTF file\n",
    "with open(output_gtf_file, \"w\") as out_file:\n",
    "    for row in gtf_df.itertuples(index=False, name=None):\n",
    "        # write the updated record to the output file\n",
    "        out_file.write(\"\\t\".join(str(x) for x in row[:8]) + \"\\t\" + row[8] + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "353851af",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "        \n",
    "with open(gtf_file, \"r\") as gtf_file:\n",
    "                    # Use GFF.parse to read the features from the file\n",
    "                    gff_handle = list(GFF.parse(gtf_file))\n",
    "#                     gff_features = gff_handle[0].features\n",
    "                    for feature in gff_handle[34].features:\n",
    "                        print(feature)\n",
    "                        # Check if this feature overlaps with the insertion site\n",
    "                        if feature.location.start.position >= insertion_site :\n",
    "                            # Calculate the new start and end positions\n",
    "                            new_start =  feature.location.start.position+ insertion_length\n",
    "                            new_end = feature.location.end.position+ insertion_length\n",
    "                            feature.location = FeatureLocation(new_start, new_end, feature.location.strand)\n",
    "\n",
    "                        if feature.location.end.position > insertion_site and  feature.location.start.position < insertion_site:\n",
    "                            # Calculate the new start and end positions\n",
    "                            new_start =  feature.location.start.position\n",
    "                            new_end =  feature.location.end.position + insertion_length\n",
    "                            \n",
    "                            # Create a new FeatureLocation object with the modified positions\n",
    "#                             modified_location = FeatureLocation(new_start, new_end, feature.features.location.strand)\n",
    "\n",
    "                            # Update the feature with the modified location\n",
    "                            feature.location = FeatureLocation(new_start, new_end, feature.location.strand)\n",
    "\n",
    "                # Create a new SeqRecord object with the modified features\n",
    "# modified_seq_record = SeqRecord(seq_record.seq, id=seq_record.id, name=seq_record.name, description=seq_record.description, features=modified_features)\n",
    "\n",
    "                # Write the modified sequence to a new GTF file\n",
    "with open(output_gtf_file, \"w\") as output_file:\n",
    "                    GFF.write(gff_handle, output_file)\n",
    "#         else:\n",
    "#             # If this is not the modified sequence, copy it to the new GTF file unmodified\n",
    "#             with open(\"mm10.gtf\", \"r\") as gtf_file:\n",
    "#                 with open(\"modified_mm10.gtf\", \"a\") as output_file:\n",
    "#                     output_file.write(gtf_file.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a9df20ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_gtf_file,'r') as gtf_file:\n",
    "    records = list(GFF.parse(gtf_file))\n",
    "\n",
    "for record in records:\n",
    "    for gene in record.features:\n",
    "        if gene.type == 'gene':\n",
    "            for transcript in gene.sub_features:\n",
    "                if transcript.type == 'transcript':\n",
    "                    for exon in transcript.sub_features:\n",
    "                        if exon.type == 'exon':\n",
    "                            # Do something with the exon feature\n",
    "                            print(exon)\n",
    "                            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f788ce3f",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'set' object has no attribute 'items'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 5\u001b[0m\n\u001b[1;32m      1\u001b[0m output_gtf_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmm10_genome/mm10-2020-A-reference-sources/gencode.vM23.primary_assembly.annotation.gtf\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(output_gtf_file, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m gtf_file:\n\u001b[1;32m      4\u001b[0m                     \u001b[38;5;66;03m# Use GFF.parse to read the features from the file\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m                     gff_handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mGFF\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgtf_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43mlimit_info\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mexon\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m#                     gff_features = gff_handle[0].features\u001b[39;00m\n\u001b[1;32m      7\u001b[0m                     \u001b[38;5;28;01mfor\u001b[39;00m feature \u001b[38;5;129;01min\u001b[39;00m gff_handle[\u001b[38;5;241m34\u001b[39m]\u001b[38;5;241m.\u001b[39mfeatures:\n",
      "File \u001b[0;32m~/miniconda3/envs/sc/lib/python3.10/site-packages/BCBio/GFF/GFFParser.py:793\u001b[0m, in \u001b[0;36mparse\u001b[0;34m(gff_files, base_dict, limit_info, target_lines)\u001b[0m\n\u001b[1;32m    790\u001b[0m \u001b[38;5;124;03m\"\"\"High level interface to parse GFF files into SeqRecords and SeqFeatures.\u001b[39;00m\n\u001b[1;32m    791\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    792\u001b[0m parser \u001b[38;5;241m=\u001b[39m GFFParser()\n\u001b[0;32m--> 793\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m rec \u001b[38;5;129;01min\u001b[39;00m parser\u001b[38;5;241m.\u001b[39mparse_in_parts(gff_files, base_dict, limit_info,\n\u001b[1;32m    794\u001b[0m         target_lines):\n\u001b[1;32m    795\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m rec\n",
      "File \u001b[0;32m~/miniconda3/envs/sc/lib/python3.10/site-packages/BCBio/GFF/GFFParser.py:332\u001b[0m, in \u001b[0;36m_AbstractMapReduceGFF.parse_in_parts\u001b[0;34m(self, gff_files, base_dict, limit_info, target_lines)\u001b[0m\n\u001b[1;32m    324\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mparse_in_parts\u001b[39m(\u001b[38;5;28mself\u001b[39m, gff_files, base_dict\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, limit_info\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    325\u001b[0m         target_lines\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    326\u001b[0m     \u001b[38;5;124;03m\"\"\"Parse a region of a GFF file specified, returning info as generated.\u001b[39;00m\n\u001b[1;32m    327\u001b[0m \n\u001b[1;32m    328\u001b[0m \u001b[38;5;124;03m    target_lines -- The number of lines in the file which should be used\u001b[39;00m\n\u001b[1;32m    329\u001b[0m \u001b[38;5;124;03m    for each partial parse. This should be determined based on available\u001b[39;00m\n\u001b[1;32m    330\u001b[0m \u001b[38;5;124;03m    memory.\u001b[39;00m\n\u001b[1;32m    331\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 332\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m results \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparse_simple(gff_files, limit_info, target_lines):\n\u001b[1;32m    333\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m base_dict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    334\u001b[0m             cur_dict \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m()\n",
      "File \u001b[0;32m~/miniconda3/envs/sc/lib/python3.10/site-packages/BCBio/GFF/GFFParser.py:352\u001b[0m, in \u001b[0;36m_AbstractMapReduceGFF.parse_simple\u001b[0;34m(self, gff_files, limit_info, target_lines)\u001b[0m\n\u001b[1;32m    350\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(gff_files, (\u001b[38;5;28mlist\u001b[39m, \u001b[38;5;28mtuple\u001b[39m)):\n\u001b[1;32m    351\u001b[0m     gff_files \u001b[38;5;241m=\u001b[39m [gff_files]\n\u001b[0;32m--> 352\u001b[0m limit_info \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_normalize_limit_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlimit_info\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    353\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m results \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_gff_process(gff_files, limit_info, target_lines):\n\u001b[1;32m    354\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m results\n",
      "File \u001b[0;32m~/miniconda3/envs/sc/lib/python3.10/site-packages/BCBio/GFF/GFFParser.py:361\u001b[0m, in \u001b[0;36m_AbstractMapReduceGFF._normalize_limit_info\u001b[0;34m(self, limit_info)\u001b[0m\n\u001b[1;32m    359\u001b[0m final_limit_info \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m    360\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m limit_info:\n\u001b[0;32m--> 361\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m key, values \u001b[38;5;129;01min\u001b[39;00m \u001b[43mlimit_info\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m():\n\u001b[1;32m    362\u001b[0m         final_limit_info[key] \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    363\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m v \u001b[38;5;129;01min\u001b[39;00m values:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'set' object has no attribute 'items'"
     ]
    }
   ],
   "source": [
    "output_gtf_file = \"mm10_genome/mm10-2020-A-reference-sources/gencode.vM23.primary_assembly.annotation.gtf\"\n",
    "\n",
    "with open(output_gtf_file, \"r\") as gtf_file:\n",
    "                    # Use GFF.parse to read the features from the file\n",
    "                    gff_handle = list(GFF.parse(gtf_file,limit_info={\"exon\":0}))\n",
    "#                     gff_features = gff_handle[0].features\n",
    "                    for feature in gff_handle[34].features:\n",
    "                        print(feature.type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef50d0be",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'updated_mm10.gtf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mupdated_mm10.gtf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m gtf_file:\n\u001b[1;32m      2\u001b[0m                     \u001b[38;5;66;03m# Use GFF.parse to read the features from the file\u001b[39;00m\n\u001b[1;32m      3\u001b[0m                     gff_handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(GFF\u001b[38;5;241m.\u001b[39mparse(gtf_file))\n",
      "File \u001b[0;32m~/miniconda3/envs/sc/lib/python3.10/site-packages/IPython/core/interactiveshell.py:282\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    275\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m}:\n\u001b[1;32m    276\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    277\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIPython won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt let you open fd=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m by default \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    278\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    279\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou can use builtins\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m open.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    280\u001b[0m     )\n\u001b[0;32m--> 282\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mio_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'updated_mm10.gtf'"
     ]
    }
   ],
   "source": [
    "\n",
    "with open('updated_mm10.gtf', \"r\") as gtf_file:\n",
    "                    # Use GFF.parse to read the features from the file\n",
    "                    gff_handle = list(GFF.parse(gtf_file))\n",
    "# #                     gff_features = gff_handle[0].features\n",
    "#                     for feature in gff_handle[0].features:\n",
    "#                         print(feature.location.end)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "edd19131",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: gene\n",
      "location: [9123:58882](-)\n",
      "qualifiers:\n",
      "    Key: gene_id, Value: ['ENSMUSG00000079800.2']\n",
      "    Key: gene_name, Value: ['AC125149.3']\n",
      "    Key: gene_type, Value: ['protein_coding']\n",
      "    Key: level, Value: ['3']\n",
      "    Key: source, Value: ['ENSEMBL']\n",
      "\n",
      "type: gene\n",
      "location: [108389:110303](-)\n",
      "qualifiers:\n",
      "    Key: gene_id, Value: ['ENSMUSG00000095092.1']\n",
      "    Key: gene_name, Value: ['AC125149.5']\n",
      "    Key: gene_type, Value: ['protein_coding']\n",
      "    Key: level, Value: ['3']\n",
      "    Key: source, Value: ['ENSEMBL']\n",
      "\n",
      "type: gene\n",
      "location: [123791:124928](+)\n",
      "qualifiers:\n",
      "    Key: gene_id, Value: ['ENSMUSG00000079192.2']\n",
      "    Key: gene_name, Value: ['AC125149.1']\n",
      "    Key: gene_type, Value: ['protein_coding']\n",
      "    Key: level, Value: ['3']\n",
      "    Key: source, Value: ['ENSEMBL']\n",
      "\n",
      "type: gene\n",
      "location: [135394:136519](-)\n",
      "qualifiers:\n",
      "    Key: gene_id, Value: ['ENSMUSG00000079794.2']\n",
      "    Key: gene_name, Value: ['AC125149.2']\n",
      "    Key: gene_type, Value: ['protein_coding']\n",
      "    Key: level, Value: ['3']\n",
      "    Key: source, Value: ['ENSEMBL']\n",
      "\n",
      "type: gene\n",
      "location: [147791:149707](+)\n",
      "qualifiers:\n",
      "    Key: gene_id, Value: ['ENSMUSG00000094799.1']\n",
      "    Key: gene_name, Value: ['AC125149.4']\n",
      "    Key: gene_type, Value: ['protein_coding']\n",
      "    Key: level, Value: ['3']\n",
      "    Key: source, Value: ['ENSEMBL']\n",
      "\n",
      "type: inferred_parent\n",
      "location: [9123:58882](-)\n",
      "id: ENSMUST00000096862.4\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['ENSMUST00000096862.4']\n",
      "\n",
      "type: inferred_parent\n",
      "location: [108389:110303](-)\n",
      "id: ENSMUST00000177562.1\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['ENSMUST00000177562.1']\n",
      "\n",
      "type: inferred_parent\n",
      "location: [123791:124928](+)\n",
      "id: ENSMUST00000111364.2\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['ENSMUST00000111364.2']\n",
      "\n",
      "type: inferred_parent\n",
      "location: [135394:136519](-)\n",
      "id: ENSMUST00000115931.2\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['ENSMUST00000115931.2']\n",
      "\n",
      "type: inferred_parent\n",
      "location: [147791:149707](+)\n",
      "id: ENSMUST00000115928.2\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['ENSMUST00000115928.2']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for feature in gff_handle[0].features:\n",
    "                        print(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5aca7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Open the mm10 FASTA file\n",
    "with open(\"mm10_genome/mm10-2020-A-reference-sources/Mus_musculus.GRCm38.dna.primary_assembly.fa\", \"r\") as fasta_file:\n",
    "    # Use SeqIO.parse to read the sequences from the file\n",
    "    sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "    # Create an empty list to store the modified sequences\n",
    "    modified_sequences = []\n",
    "\n",
    "    # Loop through each sequence in the file\n",
    "    for seq_record in sequences:\n",
    "        # Check if the subsequence is in the sequence\n",
    "        if \"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\" in str(seq_record.seq):\n",
    "            # Create a new sequence object with the modified subsequence\n",
    "            old_sequence = seq_record.seq\n",
    "            new_sequence = Seq(str(old_sequence).replace(seq, new_seq))\n",
    "\n",
    "            # Create a new SeqRecord object with the modified sequence\n",
    "            modified_seq_record = seq_record\n",
    "            modified_seq_record.seq = new_sequence\n",
    "\n",
    "            # Add the modified sequence to the list of modified sequences\n",
    "            modified_sequences.append(modified_seq_record)\n",
    "            print(seq_record.id)\n",
    "        else:\n",
    "            # Add the unmodified sequence to the list of modified sequences\n",
    "            modified_sequences.append(seq_record)\n",
    "\n",
    "# Write the modified sequences to a new FASTA file\n",
    "with open(\"modified_mm10.fa\", \"w\") as output_file:\n",
    "    SeqIO.write(modified_sequences, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2343b33c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "woow\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "new_seq=\"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTCCCCCTCCCCCGACTCGATTGTACGTGAAGCTCCCCCTCCCCCGATCAAGCACACAAAAAACCAACACACAGATCTAATGAAAATAAAGATCTTTTATTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\"\n",
    "\n",
    "# Open the mm10 FASTA file\n",
    "with open(\"updated_mm10.fa\", \"r\") as fasta_file:\n",
    "    # Use SeqIO.parse to read the sequences from the file\n",
    "    sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "    # Create an empty list to store the modified sequences\n",
    "\n",
    "    # Loop through each sequence in the file\n",
    "    for seq_record in sequences:\n",
    "        # Check if the subsequence is in the sequence\n",
    "#         if \"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\" in str(seq_record.seq):\n",
    "        if new_seq in str(seq_record.seq):\n",
    "            print('woow')\n",
    "            print(seq_record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a87eae03",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = new_seq\n",
    "# Open the FASTA file and search for the sequence\n",
    "with open(\"modified_mm10.fa\", \"r\") as f:\n",
    "    header, seq = None, \"\"\n",
    "    for line in f:\n",
    "        if line.startswith(\">\"):\n",
    "            if seq.upper().find(sequence) != -1:\n",
    "                print(header)\n",
    "            header = line.strip()\n",
    "            seq=\"\"\n",
    "        else:\n",
    "            if header!= None:\n",
    "                seq += line.strip()\n",
    "#         if seq.upper().find(sequence) != -1:\n",
    "#             print(header)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "644be589",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence[105100761,105100790]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "597ed5bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = \"TGTTGCAGGTTGGGAAGGCGAGTCAGTGGGCAGCATTCCCATCTCCAGCTTGTACAAATCCCTGCATCCCATCTCCAGCCCTGCCGAAACCAGGCAGGGTAGCTCCTGTTTTGTAATCCTGCTGCTCAGGAAGTGGAGGCAGGAGGATCGGAAGTTCAAGGTCATTCCAGTC\"\n",
    "\n",
    "if seq.upper().find(sequence) != -1:\n",
    "            print(header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942d24ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Open the FASTA file and search for the sequence\n",
    "with open(genome_file, \"r\") as f:\n",
    "    header, seq = None, \"\"\n",
    "    for line in f:\n",
    "        line.strip()\n",
    "        if line.startswith(\">\"):\n",
    "            print(line.strip())\n",
    "            if seq.upper().find(sequence) != -1:\n",
    "                print(header)\n",
    "                break\n",
    "            header, seq = line.strip(), \"\"\n",
    "        else:\n",
    "            seq += line.strip()\n",
    "    if seq.upper().find(sequence) != -1:\n",
    "        print(header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e4e335d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc",
   "language": "python",
   "name": "sc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
