{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12aae311",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load libraries\n",
    "library(Signac)\n",
    "library(GenomicRanges)\n",
    "library(future)\n",
    "library(EnsDb.Mmusculus.v79)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "#\n",
    "\n",
    "#Load peak data as GRanges\n",
    "sampleNames<-c(\"ctrl1\",\"ctrl2\",\"ko1\",\"ko2\")\n",
    "sample_loc<-\"new_h5/\"\n",
    "peaks<-lapply(sampleNames,function(x){makeGRangesFromDataFrame(read.table(\n",
    "      file=paste0(sample_loc,x,\"/atac_peaks.bed\"),\n",
    "      col.names = c(\"chr\", \"start\", \"end\")\n",
    "      ))}\n",
    "      )\n",
    "## Create a unified set of peaks to quantify in each dataset\n",
    "combined.peaks <- reduce(x = c(peaks[[1]],peaks[[2]],peaks[[3]],peaks[[4]])) #There's probably a better way to do this\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58cd1f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out bad peaks based on length if necessary.\n",
    "peakwidths <- width(combined.peaks)\n",
    "summary(peakwidths) # Not for this dataset\n",
    "\n",
    "#Create fragment objects\n",
    "md<-lapply(sampleNames,function(x)read.table(\n",
    "                                  paste0(sample_loc,x,\"/filtered_feature_bc_matrix/barcodes.tsv.gz\"))$V1\n",
    "                                  )\n",
    "names(md)<-sampleNames\n",
    "\n",
    "frags<-lapply(sampleNames,function(x)CreateFragmentObject(\n",
    "  paste0(sample_loc,x,\"/atac_fragments.tsv.gz\")\n",
    "  ,cells = md[[x]])\n",
    "  )\n",
    "names(frags)<-sampleNames\n",
    "\n",
    "#Quantify peaks in each dataset\n",
    "peaks.counts<-lapply(names(frags), function(frag)FeatureMatrix(fragments=frags[[frag]],features=combined.peaks,cells=md[[frag]]))\n",
    "names(peaks.counts)<-sampleNames\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8c32ad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get annotation for mm10\n",
    "annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "seqlevelsStyle(annotation) <- \"UCSC\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47cfed55",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# set up sample loop to load the RNA data,merge with ATAC and save to seurat object\n",
    "samples<-list()\n",
    "for(i in sampleNames){\n",
    "  setwd(paste0(sample_loc,i,\"/\"))\n",
    "  counts <- Read10X_h5(\"filtered_feature_bc_matrix.h5\") #count data\n",
    "\n",
    "  # create a Seurat object containing the RNA adata\n",
    "  dat <- CreateSeuratObject(\n",
    "    counts = counts$`Gene Expression`,\n",
    "    assay = \"RNA\"\n",
    "  )\n",
    "  \n",
    "  # create ATAC assay and add it to the object\n",
    "  dat[[\"ATAC\"]] <- CreateChromatinAssay(\n",
    "    counts = peaks.counts[[i]],\n",
    "    sep = c(\":\", \"-\"),\n",
    "    fragments = frags[[i]],\n",
    "    annotation = annotation\n",
    "  )\n",
    "  samples[[i]]<-dat\n",
    "}\n",
    "#Merge all in one seurat object\n",
    "\n",
    "combined <- merge(\n",
    "  x = samples[[1]],\n",
    "  y = samples[-1],\n",
    "  add.cell.ids = c(\"Ctrl1\", \"Ctrl2\", \"KO1\", \"KO2\")\n",
    ")\n",
    "\n",
    "#ATAC QC\n",
    "DefaultAssay(combined) <- \"ATAC\"\n",
    "\n",
    "combined <- NucleosomeSignal(combined)\n",
    "combined <- TSSEnrichment(combined)\n",
    "\n",
    "VlnPlot(\n",
    "  object = combined,\n",
    "  features = c(\"nCount_RNA\", \"nCount_ATAC\", \"TSS.enrichment\", \"nucleosome_signal\"),\n",
    "  ncol = 4,\n",
    "  pt.size = 0\n",
    ")\n",
    "\n",
    "# The filtering thresholds are arbitrary; better approach is probably necessary\n",
    "combined <- subset(\n",
    "  x = combined,\n",
    "  subset = nCount_ATAC < 50000 &\n",
    "    nCount_RNA < 25000 &\n",
    "    nCount_ATAC > 1000 &\n",
    "    nCount_RNA > 1000 &\n",
    "    nucleosome_signal < 2 &\n",
    "    TSS.enrichment > 1\n",
    ")\n",
    "\n",
    "#Standard multiome normalization/dr analysis\n",
    "\n",
    "DefaultAssay(combined) <- \"RNA\"\n",
    "combined <- SCTransform(combined)\n",
    "combined <- RunPCA(combined)\n",
    "\n",
    "combined <- RunTFIDF(combined)\n",
    "combined <- FindTopFeatures(combined, min.cutoff = 20)\n",
    "combined <- RunSVD(combined)\n",
    "combined <- RunUMAP(combined, dims = 1:30, reduction = 'lsi')\n",
    "\n",
    "#Cell type assignment from reference\n",
    "library(scRNAseq)\n",
    "library(SeuratDisk)\n",
    "library(zellkonverter)\n",
    "reference1<-readH5AD(\"/data/user/apapada1/Downloads/kidney.h5ad\")\n",
    "meta<-as.data.frame(reference1@colData)\n",
    "reference<-CreateSeuratObject(counts=assay(reference1,\"X\"),meta.data = as.data.frame(colData(reference1)))\n",
    "reference<-SCTransform(reference)\n",
    "reference <- RunPCA(reference)\n",
    "reference <- RunUMAP(reference, dims = 1:30)\n",
    "DimPlot(combined, group.by = 'dataset', pt.size = 0.1,reduction=\"lsi\")\n",
    "DimPlot(reference, group.by = 'cell_type', pt.size = 0.1)\n",
    "\n",
    "#Anchor transfer and predictions\n",
    "transfer_anchors <- FindTransferAnchors(\n",
    "  reference = reference1,\n",
    "  query = combined,\n",
    "  normalization.method = \"LogNormalize\",\n",
    "  recompute.residuals = FALSE,\n",
    "  dims = 1:30)\n",
    "predictions <- TransferData(\n",
    "  anchorset = transfer_anchors, \n",
    "  refdata = reference1$type,\n",
    "  weight.reduction = combined[['pca']],\n",
    "  dims = 1:30\n",
    ")\n",
    "combined <- AddMetaData(\n",
    "  object = combined,\n",
    "  metadata = predictions\n",
    ")# You can filter based on prediction score here.\n",
    "\n",
    "#Multimodal umap\n",
    "\n",
    "combined<- FindMultiModalNeighbors(\n",
    "  object = combined,\n",
    "  reduction.list = list(\"pca\", \"lsi\"), \n",
    "  dims.list = list(1:30, 2:40),\n",
    "  modality.weight.name = \"RNA.weight\",\n",
    "  verbose = TRUE\n",
    ")\n",
    "combined <- RunUMAP(\n",
    "  object = combined,\n",
    "  nn.name = \"weighted.nn\",\n",
    "  assay = \"RNA\",\n",
    "  verbose = TRUE\n",
    ")\n",
    "\n",
    "#Keep main chromosome peaks\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "main.chroms <- standardChromosomes(BSgenome.Mmusculus.UCSC.mm10)\n",
    "keep.peaks <- as.logical(seqnames(granges(combined)) %in% main.chroms)\n",
    "combined[[\"ATAC\"]] <- subset(combined[[\"ATAC\"]], features = rownames(combined[[\"ATAC\"]])[keep.peaks])\n",
    "\n",
    "#Coverage plots\n",
    "\n",
    "combined <- LinkPeaks(\n",
    "  object = combined,\n",
    "  peak.assay = \"ATAC\",\n",
    "  expression.assay = \"SCT\",\n",
    "  genes.use = c(\"Nphs1\", \"Wt1\")\n",
    ")\n",
    "Idents(combined) <- \"predicted.id\"\n",
    "p1 <- CoveragePlot(\n",
    "  object = combined,\n",
    "  region = \"Nphs1\",\n",
    "  features = \"Nphs1\",\n",
    "  expression.assay = \"SCT\",\n",
    "  extend.upstream = 500,\n",
    "  extend.downstream = 10000\n",
    ")\n",
    "p2 <- CoveragePlot(\n",
    "  object = combined,\n",
    "  region = \"Wt1\",\n",
    "  features = \"Wt1\",\n",
    "  expression.assay = \"SCT\",\n",
    "  extend.upstream = 500,\n",
    "  extend.downstream = 10000\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599401f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
