{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bccc7c6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is an example script for the standard KFO single-cell/single-nuclei data analysis. \n",
    "#This specific example is ran on 10x single-nuclei data from glom prep.\n",
    "\n",
    "## Set seed for replicability\n",
    "set.seed(1234)\n",
    "\n",
    "##Load libraries\n",
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(celda)\n",
    "library(scater)\n",
    "library(scDblFinder)\n",
    "library(gridExtra)\n",
    "library(writexl)\n",
    "library(gprofiler2)\n",
    "library(decontX)\n",
    "\n",
    "##Folders\n",
    "\n",
    "dir= \"/data/public/kbohl1/Mai13_TXM04_Michael/cellranger\"\n",
    "seu_path <- \"./data/Seurat Objects/\"\n",
    "fig_out <- \"./figures/\"\n",
    "data_out <- \"./data/\"\n",
    "res_out  <- \"./results/\"\n",
    "dir.create(fig_out)\n",
    "dir.create(data_out)\n",
    "dir.create(res_out)\n",
    "dir.create(seu_path)\n",
    "\n",
    "#Note: In the comments, library and cell are used interchangeably, even though, especially in droplet-based data, they are different.\n",
    "#While a cell barcode should correspond to one cell, we have to correct for empty droplets and doublets.\n",
    "\n",
    "##Step 1: Loading the data and removing ambient RNA\n",
    "##The following custom function loads data from CellRanger.\n",
    "\n",
    "#' Load in data from 10X\n",
    "#'\n",
    "#' Enables easy loading of sparse data matrices provided by 10X genomics.\n",
    "#'\n",
    "#' @param data.dir Directory containing the matrix.mtx, genes.tsv (or features.tsv), and barcodes.tsv\n",
    "#' files provided by 10X. A vector or named vector can be given in order to load\n",
    "#' several data directories. If a named vector is given, the cell barcode names\n",
    "#' will be prefixed with the name.\n",
    "#' @param gene.column Specify which column of genes.tsv or features.tsv to use for gene names; default is 2\n",
    "#' @param cell.column Specify which column of barcodes.tsv to use for cell names; default is 1\n",
    "#' @param unique.features Make feature names unique (default TRUE)\n",
    "#' @param strip.suffix Remove trailing \"-1\" if present in all cell barcodes.\n",
    "#'\n",
    "#' @return If features.csv indicates the data has multiple data types, a list\n",
    "#'   containing a sparse matrix of the data from each type will be returned.\n",
    "#'   Otherwise a sparse matrix containing the expression data will be returned.\n",
    "#'\n",
    "#' @importFrom Matrix readMM\n",
    "#' @importFrom utils read.delim\n",
    "#'\n",
    "#' @export\n",
    "#' @concept preprocessing\n",
    "#'\n",
    "#' @examples\n",
    "#' \\dontrun{\n",
    "#' # For output from CellRanger < 3.0\n",
    "#' data_dir <- 'path/to/data/directory'\n",
    "#' list.files(data_dir) # Should show barcodes.tsv, genes.tsv, and matrix.mtx\n",
    "#' expression_matrix <- Read10X(data.dir = data_dir)\n",
    "#' seurat_object = CreateSeuratObject(counts = expression_matrix)\n",
    "#'\n",
    "#' # For output from CellRanger >= 3.0 with multiple data types\n",
    "#' data_dir <- 'path/to/data/directory'\n",
    "#' list.files(data_dir) # Should show barcodes.tsv.gz, features.tsv.gz, and matrix.mtx.gz\n",
    "#' data <- Read10X(data.dir = data_dir)\n",
    "#' seurat_object = CreateSeuratObject(counts = data$`Gene Expression`)\n",
    "#' seurat_object[['Protein']] = CreateAssayObject(counts = data$`Antibody Capture`)\n",
    "#' }\n",
    "#'\n",
    "Read10X <- function(\n",
    "    data.dir = NULL,\n",
    "    gene.column = 2,\n",
    "    cell.column = 1,\n",
    "    unique.features = TRUE,\n",
    "    strip.suffix = FALSE, \n",
    "    zipped=F \n",
    ") \n",
    "{\n",
    "  full.data <- list()\n",
    "  for (i in seq_along(along.with = data.dir)) {\n",
    "    run <- data.dir[i]\n",
    "    if (!dir.exists(paths = run)) {\n",
    "      stop(\"Directory provided does not exist\")\n",
    "    }\n",
    "    barcode.loc <- file.path(run, 'barcodes.tsv')\n",
    "    gene.loc <- file.path(run, 'genes.tsv')\n",
    "    features.loc <- file.path(run, 'features.tsv')\n",
    "    matrix.loc <- file.path(run, 'matrix.mtx')\n",
    "    # Flag to indicate if this data is from CellRanger >= 3.0\n",
    "    pre_ver_3 <- file.exists(gene.loc)\n",
    "    if (!pre_ver_3 & zipped==T ) {\n",
    "      addgz <- function(s) {\n",
    "        return(paste0(s, \".gz\"))\n",
    "      }\n",
    "      barcode.loc <- addgz(s = barcode.loc)\n",
    "      matrix.loc <- addgz(s = matrix.loc)\n",
    "      features.loc <- addgz(s = features.loc)\n",
    "    }\n",
    "    if (!file.exists(barcode.loc)) {\n",
    "      stop(\"Barcode file missing. Expecting \", basename(path = barcode.loc))\n",
    "    }\n",
    "    if (!pre_ver_3 && !file.exists(features.loc) ) {\n",
    "      stop(\"Gene name or features file missing. Expecting \", basename(path = features.loc))\n",
    "    }\n",
    "    if (!file.exists(matrix.loc)) {\n",
    "      stop(\"Expression matrix file missing. Expecting \", basename(path = matrix.loc))\n",
    "    }\n",
    "    data <- readMM(file = matrix.loc)\n",
    "    cell.barcodes <- read.table(file = barcode.loc, header = FALSE, sep = '\\t', row.names = NULL)\n",
    "    if (ncol(x = cell.barcodes) > 1) {\n",
    "      cell.names <- cell.barcodes[, cell.column]\n",
    "    } else {\n",
    "      cell.names <- readLines(con = barcode.loc)\n",
    "    }\n",
    "    if (all(grepl(pattern = \"\\\\-1$\", x = cell.names)) & strip.suffix) {\n",
    "      cell.names <- as.vector(x = as.character(x = sapply(\n",
    "        X = cell.names,\n",
    "        FUN = ExtractField,\n",
    "        field = 1,\n",
    "        delim = \"-\"\n",
    "      )))\n",
    "    }\n",
    "    if (is.null(x = names(x = data.dir))) {\n",
    "      if (i < 2) {\n",
    "        colnames(x = data) <- cell.names\n",
    "      } else {\n",
    "        colnames(x = data) <- paste0(i, \"_\", cell.names)\n",
    "      }\n",
    "    } else {\n",
    "      colnames(x = data) <- paste0(names(x = data.dir)[i], \"_\", cell.names)\n",
    "    }\n",
    "    feature.names <- read.delim(\n",
    "      file = ifelse(test = pre_ver_3, yes = gene.loc, no = features.loc),\n",
    "      header = FALSE,\n",
    "      stringsAsFactors = FALSE\n",
    "    )\n",
    "    if (any(is.na(x = feature.names[, gene.column]))) {\n",
    "      warning(\n",
    "        'Some features names are NA. Replacing NA names with ID from the opposite column requested',\n",
    "        call. = FALSE,\n",
    "        immediate. = TRUE\n",
    "      )\n",
    "      na.features <- which(x = is.na(x = feature.names[, gene.column]))\n",
    "      replacement.column <- ifelse(test = gene.column == 2, yes = 1, no = 2)\n",
    "      feature.names[na.features, gene.column] <- feature.names[na.features, replacement.column]\n",
    "    }\n",
    "    if (unique.features) {\n",
    "      fcols = ncol(x = feature.names)\n",
    "      if (fcols < gene.column) {\n",
    "        stop(paste0(\"gene.column was set to \", gene.column,\n",
    "                    \" but feature.tsv.gz (or genes.tsv) only has \", fcols, \" columns.\",\n",
    "                    \" Try setting the gene.column argument to a value <= to \", fcols, \".\"))\n",
    "      }\n",
    "      rownames(x = data) <- make.unique(names = feature.names[, gene.column])\n",
    "    }\n",
    "    # In cell ranger 3.0, a third column specifying the type of data was added\n",
    "    # and we will return each type of data as a separate matrix\n",
    "    if (ncol(x = feature.names) > 2) {\n",
    "      data_types <- factor(x = feature.names$V3)\n",
    "      lvls <- levels(x = data_types)\n",
    "      if (length(x = lvls) > 1 && length(x = full.data) == 0) {\n",
    "        message(\"10X data contains more than one type and is being returned as a list containing matrices of each type.\")\n",
    "      }\n",
    "      expr_name <- \"Gene Expression\"\n",
    "      if (expr_name %in% lvls) { # Return Gene Expression first\n",
    "        lvls <- c(expr_name, lvls[-which(x = lvls == expr_name)])\n",
    "      }\n",
    "      data <- lapply(\n",
    "        X = lvls,\n",
    "        FUN = function(l) {\n",
    "          return(data[data_types == l, , drop = FALSE])\n",
    "        }\n",
    "      )\n",
    "      names(x = data) <- lvls\n",
    "    } else{\n",
    "      data <- list(data)\n",
    "    }\n",
    "    full.data[[length(x = full.data) + 1]] <- data\n",
    "  }\n",
    "  # Combine all the data from different directories into one big matrix, note this\n",
    "  # assumes that all data directories essentially have the same features files\n",
    "  list_of_data <- list()\n",
    "  for (j in 1:length(x = full.data[[1]])) {\n",
    "    list_of_data[[j]] <- do.call(cbind, lapply(X = full.data, FUN = `[[`, j))\n",
    "    # Fix for Issue #913\n",
    "    list_of_data[[j]] <- as(object = list_of_data[[j]], Class = \"dgCMatrix\")\n",
    "  }\n",
    "  names(x = list_of_data) <- names(x = full.data[[1]])\n",
    "  # If multiple features, will return a list, otherwise\n",
    "  # a matrix.\n",
    "  if (length(x = list_of_data) == 1) {\n",
    "    return(list_of_data[[1]])\n",
    "  } else {\n",
    "    return(list_of_data)\n",
    "  }\n",
    "}\n",
    "\n",
    "##List 10x folders. This step can be different, depending on your data.\n",
    "lldir <- list.dirs(dir, recursive = F)\n",
    "\n",
    "##Get vector of sample names. Here we get them from the folder names, but it can be done manually, from annotation files etc.\n",
    "sample_names <- basename(lldir)\n",
    "\n",
    "## Run DecontX on each individual mouse dataset. Ambient RNA removal is optional but empirically it improves our dataset quality.\n",
    "\n",
    "for(i in 1:length(sample_names)){\n",
    "  print(sample_names[i])\n",
    "#Read filtered data\n",
    "  sce <- Read10X(data.dir = paste0(lldir[i] , \"/outs/filtered_feature_bc_matrix/\"),  gene.column =2, zipped=T)\n",
    "  sce<-sce$`Gene Expression` #If multiome.\n",
    "#Read raw data\n",
    "  sce.raw <- Read10X(data.dir = paste0(lldir[i] , \"/outs/raw_feature_bc_matrix/\"), gene.column =2, zipped=T)\n",
    "  sce.raw<-sce.raw$`Gene Expression`  #If multiome.\n",
    "#Use decontX to remove ambient RNA  \n",
    "  sce.decont <- decontX(sce, background = sce.raw)\n",
    "#Save file after ambient RNA removal\n",
    "  saveRDS(sce.decont , paste0(seu_path , \"filtered_decontX.\",sample_names[i], \".rda\"))\n",
    "  rm(sce, sce.decont, sce.raw)\n",
    "  print(\" \")\n",
    "}\n",
    "\n",
    "##Create a list of Seurat objects including all data.\n",
    "\n",
    "seuList <- lapply(seq(lldir), function(i){\n",
    "  print(i)\n",
    "  # Read counts after ambient RNA removal\n",
    "  cmat.decontX <- readRDS(paste0(data_out, \"filtered_decontX.\", sample_names[i], \".rda\"))\n",
    "  # Create seurat object\n",
    "  ssc.decontX <- CreateSeuratObject(counts = round(cmat.decontX$decontXcounts))\n",
    "  return(ssc.decontX)\n",
    "})\n",
    "\n",
    "##Merge all samples in one object\n",
    "\n",
    "seu <- merge(seuList[[1]] , y= seuList[2:length(seuList)] ,\n",
    "              add.cell.ids= sample_names)\n",
    "\n",
    "##Add metadata \n",
    "\n",
    "seu$sample<-sub( \"_.*\" , \"\" , colnames(seu)) #Sample labels\n",
    "\n",
    "#More metadata columns can be added if relevant to the project(e.g. age, technical batch, condition, sex)\n",
    "\n",
    "##Step :Doublet Identification\n",
    "\n",
    "##Introduction\n",
    "\n",
    "#Doublets, defined as two cells sequenced under the same cellular barcode (e.g., being captured in the same droplet), are fairly frequent in scRNAseq datasets,\n",
    "#with estimates ranging from 1 to 10% depending on the platform and cell concentration used. While doublets of the same cell type are relatively innocuous for \n",
    "#most downstream applications due to their conservation of the relative expression between genes, doublets formed from different cell types or states are likely \n",
    "#to be misclassified and could potentially distort downstream analysis.\n",
    "\n",
    "#In some cases, doublets can be identified through their unusually high number of reads and detected features, but this is not always the case.\n",
    "#Therefore, a number of methods were developed to identify doublets, in most cases by comparing each cell to artificially created doublets.\n",
    "#In this script, we use scDblFinder.\n",
    "#Alternatives: DoubletFinder, doubletCell, scds\n",
    "\n",
    "#Convert to SingleCellExperiment\n",
    "seu_sce <- as.SingleCellExperiment(seu)\n",
    "\n",
    "#Run scDblFinder\n",
    "seu_sce <- scDblFinder(seu_sce, samples =  \"sample\", verbose = TRUE)\n",
    "\n",
    "# show amount of doublets identified per sample\n",
    "doublet_table <-  table(truth =  seu_sce$sample, call = seu_sce$scDblFinder.class)\n",
    "#Calculate percentage\n",
    "doublet_table <- as.data.table(doublet_table)\n",
    "doublet_table <- pivot_wider(doublet_table, names_from = call, values_from = N)\n",
    "doublet_table$doublet_percentage <- doublet_table$doublet/(doublet_table$singlet+doublet_table$doublet)\n",
    "#Transform back to Seurat\n",
    "seu <- as.Seurat(seu_sce, project = \"orig.ident\")\n",
    "\n",
    "#Visualize doublet score distribution\n",
    "data <- data.frame(seu$scDblFinder.score, seu$scDblFinder.class)\n",
    "singlet.df <- data[data$seu.scDblFinder.class == \"singlet\",]\n",
    "doublet.df <- data[data$seu.scDblFinder.class == \"doublet\",]\n",
    "\n",
    "ggplot(data = singlet.df, aes(x = singlet.df$seu.scDblFinder.score)) + geom_histogram()\n",
    "ggplot(data = doublet.df, aes(x = doublet.df$seu.scDblFinder.score)) + geom_histogram()\n",
    "\n",
    "rm(seu_sce)\n",
    "\n",
    "# Save post Dbl identification Seurat\n",
    "saveRDS(seu, paste0(seu_path,  \"seu_dbl.rds\"))\n",
    "\n",
    "#Step 3: Quality control\n",
    "\n",
    "##3.1 Introduction\n",
    "\n",
    "#Low quality cells can arise for multiple reasons, including cell damage and failure in library preparation.\n",
    "#Their characteristics include low RNA counts, few expressed genes and high mitochondrial/ribosomal RNA percentage.\n",
    "#These cells need to be removed from further analysis because:\n",
    "  #1)They tend to cluster together and form distinct cell subpopulations, even if they belong to different cell types.\n",
    "  #2)They reduce the efficiency of dimensional reduction, since they affect the differences captured by the first principal components.\n",
    "  #3)They affect differential expression analysis (inclusion of mitochondrial genes, overrepresentation of contaminating transcripts)\n",
    "#The removal of these cells is commonly referred to as quality control.\n",
    "\n",
    "##3.2 Metrics\n",
    "\n",
    "#To identify low-quality cells we use:\n",
    "  #a)The library size. Cells with low library size are of low quality, since most of their RNA has been lost during one point of the preparations.\n",
    "  #b)The number of expressed genes.\n",
    "  #c)The proportion of reads mapped to the mitochondrial genome. In single-cell data, high proportions are an indication of damaged cells. In single-nuclei data, \n",
    "  #high proportions are indicative of unsuccessful cytoplasm stripping.\n",
    "\n",
    "#The first two QC metrics can be found in the metadata of the Seurat object.\n",
    "summary(seu@meta.data$nCount_RNA)   #Number of RNA molecules per cell\n",
    "summary(seu@meta.data$nFeature_RNA) #Number of genes per cell\n",
    "\n",
    "#The third can be calculated.\n",
    "seu <- PercentageFeatureSet(seu, \"^mt-\", col.name = \"percent_mito\") #If mice\n",
    "seu <- PercentageFeatureSet(seu, \"^MT-\", col.name = \"percent_mito\") #If human\n",
    "summary(seu@meta.data$percent_mito)\n",
    "\n",
    "#An important assumption is that the metrics are independent of the biological state of the cell.\n",
    "#In certain datasets, there can be high variation in RNA content or mitochondrial expression in different cell types.\n",
    "#In that case, caution should be exercised to prevent their loss.\n",
    "\n",
    "#From experience, removal of ribosomal genes robustly reduced the quality of the clustering, suggesting that they represent \n",
    "#real biological differences between subpopulations.\n",
    "\n",
    "##3.3 Low quality cell identification\n",
    "\n",
    "##a)Fixed thresholds. An arbitrary fixed quality threshold is set for our metrics. Any cell outside of these thresholds is low quality.\n",
    "##Advantages:    Simple\n",
    "##Disadvantages: Different thresholds for each protocol and biological system. Even with a lot of experience, technical issues can have a big effect on the qc metrics distribution.\n",
    "##Example:\n",
    "qc.rna<-seu@meta.data$nCount_RNA<300\n",
    "qc.gene<-seu@meta.data$nFeature_RNA<200\n",
    "qc.mito<-seu@meta.data$percent_mito<0.05\n",
    "discard <- qc.rna | qc.gene | qc.mito\n",
    "#Summarize the number of discarded cells per reason.\n",
    "DataFrame(nCount=sum(qc.rna), nFeature=sum(qc.gene),percent_mito=sum(qc.mito,na.rm=T), total=sum(discard,na.rm=T))\n",
    "\n",
    "##b)Adaptive thresholds. Under the assumption that most of our cells are high quality, we remove outliers (more than 3 MADs from the median)\n",
    "##Advantages:    The majority of the cells are not removed. Less arbitrary than approach (a).\n",
    "##Disadvantages: An important risk of excluding cells on the basis of their distance from the whole distribution of cells on some properties (e.g., library size) \n",
    "##               is that these properties tend to have different distributions across subpopulations.\n",
    "##               This can lead to strong biases against certain subpopulations .\n",
    "##               Additionally, the underlying assumption of a high-quality majority may not always be appropriate\n",
    "\n",
    "qc.rna<-isOutlier(seu@meta.data$nCount_RNA)\n",
    "qc.gene<-isOutlier(seu@meta.data$nFeature_RNA)\n",
    "qc.mito<-isOutlier(seu@meta.data$percent_mito)\n",
    "discard <- qc.rna | qc.gene | qc.mito\n",
    "#Summarize the number of discarded cells per reason.\n",
    "DataFrame(nCount=sum(qc.rna), nFeature=sum(qc.gene),percent_mito=sum(qc.mito,na.rm=T), total=sum(discard,na.rm=T))\n",
    "\n",
    "##c)Based on diagnostic plots.\n",
    "##Create correlation and feature plots of the QC metrics. Remove outliers based on the plots.\n",
    "##Advantages:Intuitive, allows better judgment of whether adaptive thresholds are appropriate, allows understanding of distribution.\n",
    "##Disadvantages: Can be ill-defined/subjective.\n",
    "##Diagnostic plots should be created regardless. \n",
    "\n",
    "p1 <- FeatureScatter(seu, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\") +\n",
    "  geom_vline(xintercept = 300, color = \"red\") + geom_hline(yintercept = 8000, color = \"red\")\n",
    "p2 <- FeatureScatter(seu, feature1 = \"nCount_RNA\", feature2 = \"percent_mito\") + \n",
    "  geom_vline(xintercept = 300, color = \"red\") + \n",
    "  geom_hline(yintercept = 1, color = \"red\")\n",
    "cowplot::plot_grid(p1,p2, ncol = 2)\n",
    "p3 <- VlnPlot(seu, features = \"nCount_RNA\", pt.size = FALSE)\n",
    "p4 <- VlnPlot(seu, features = \"nFeature_RNA\", pt.size = FALSE)\n",
    "p5 <- VlnPlot(seu, features = \"percent_mito\", pt.size = FALSE)\n",
    "cowplot::plot_grid(p3,p4,p5, ncol = 3)\n",
    "p6 <- VlnPlot(seu, features = \"nCount_RNA\", pt.size = FALSE,group.by = \"sample\")\n",
    "p7 <- VlnPlot(seu, features = \"nFeature_RNA\", pt.size = FALSE,group.by = \"sample\")\n",
    "p8 <- VlnPlot(seu, features = \"percent_mito\", pt.size = FALSE,group.by = \"sample\")\n",
    "cowplot::plot_grid(p6,p7,p8, ncol = 3)\n",
    "\n",
    "##3.4 Low quality cell filtering\n",
    "##Low quality cells can either be marked or removed.\n",
    "##Marking them defers the final decision until clustering and dimensional reduction, which can allow better judgment of which criteria \n",
    "##to use. The downside is that it shifts the burden of QC to the manual interpretation of the clusters, which is already \n",
    "##a major bottleneck in scRNA-seq data analysis. Retention of low-quality cells also compromises the accuracy of the \n",
    "##variance modelling\n",
    "\n",
    "##For standard analysis, removal is highly recommended.\n",
    "seu <- subset(seu, subset = scDblFinder.class==\"singlet\"&scDblFinder.score<0.5)\n",
    "seu <- subset(seu, subset = nFeature_RNA > 200 & nCount_RNA> 500 & percent_mito < 1& nFeature_RNA<50000)\n",
    "\n",
    "\n",
    "#Step 4: Normalization\n",
    "\n",
    "##4.1 Introduction\n",
    "##Technical differences in cDNA capture or PCR amplification efficiency create differences in sequencing coverage between libraries.\n",
    "##Normalization aims to remove these differences so that differential expression/heterogeneity are driven by biological factors.\n",
    "\n",
    "##4.2 Methods\n",
    "##Two main approaches:\n",
    "##a)Library size normalization and log-transformation.\n",
    "seu<-NormalizeData(seu)\n",
    "##b)sctransform\n",
    "seu<-SCTransform(seu)\n",
    "\n",
    "##Step 5: Feature selection and scaling\n",
    "\n",
    "seu <- seu %>%\n",
    "  FindVariableFeatures(selection.method = \"vst\", nfeatures = 2000) %>%\n",
    "  ScaleData(vars.to.regress = \"percent_mito\") #Not necessary if sctransform was used for normalization\n",
    "\n",
    "##Step 6: Dimensional reduction and clustering\n",
    "\n",
    "#PCA\n",
    "seu<-RunPCA(seu)\n",
    "DimPlot(seu)\n",
    "#Elbow plot\n",
    "ElbowPlot(seu)\n",
    "#UMAP and clustering\n",
    "seu <- seu %>% \n",
    "  FindNeighbors(dims = 1:20) %>% #number of dimensions is arbitrary but elbow plot can help with the determination\n",
    "  FindClusters(resolution = 0.2) %>% #the package clustree can be used to find the optimal clustering resolution\n",
    "  RunUMAP(dims = 1:20)\n",
    "DimPlot(seu)\n",
    "\n",
    "##Step 7: Cell type identification\n",
    "\n",
    "##Find cluster markers\n",
    "seu.cluster.markers <- FindAllMarkers(seu, \n",
    "                                      logfc.threshold = 0.25,\n",
    "                                      min.pct = 0.2,\n",
    "                                      only.pos = TRUE)\n",
    "cluster_identifier <- seu.cluster.markers %>%\n",
    "  group_by(cluster) %>%\n",
    "  slice_max(n = 5, order_by = avg_log2FC)\n",
    "\n",
    "##Manual annotation\n",
    "# Visualization of DE genes for cluster identification\n",
    "DotPlot(seu, features = c(\"Nphs1\", \"Nphs2\", \"Thsd7a\", \"Itga8\", \"Gata3\", \"Flt1\", \"Pecam1\", \"Ptprc\", \"Itgax\", \"Slc34a1\", \"Slc12a3\", \"Slc4a9\"))\n",
    "DotPlot(seu, features = subset(cluster_identifier, subset = cluster == 1)$gene)\n",
    "DotPlot(seu, features = cluster_identifier$gene)+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "\n",
    "# check qc metrics in different clusters\n",
    "VlnPlot(seu, features = \"scDblFinder.score\", pt.size = F)\n",
    "\n",
    "## Compare manual annotation with PanglaoDB markers\n",
    "# AddModuleScore with Panglao markers\n",
    "panglao_markers <- fread(\"./data/panglao_celltype_markers/panglaodb_markers.tsv\")\n",
    "colnames(panglao_markers)[3] <- \"celltype\"\n",
    "# Exclude exclusively human markers\n",
    "mm_panglao_markers <- panglao_markers[species != \"Hs\",]\n",
    "# Convert gene symbols to mouse with function from nichenetr package\n",
    "library(nichenetr)\n",
    "mm_panglao_markers$mouse_symbol <- convert_human_to_mouse_symbols(mm_panglao_markers$`official gene symbol`)\n",
    "# Exclude NAs (results from conversion to mouse; not 1:1 ortholog for every human gene available)\n",
    "mm_panglao_markers <- mm_panglao_markers[is.na(mm_panglao_markers$mouse_symbol) == FALSE, ]\n",
    "\n",
    "# Subset for kidney cell types\n",
    "kidney_markers <- mm_panglao_markers[organ == \"Kidney\",]\n",
    "\n",
    "for(type in unique(kidney_markers$celltype)){\n",
    "  tmp_markers <- mm_panglao_markers[celltype == type,]\n",
    "  seu <- AddModuleScore(seu, \n",
    "                        features = list(tmp_markers$mouse_symbol),\n",
    "                        name = paste(type, \"enriched\", sep = \"_\"))\n",
    "}\n",
    "\n",
    "FeaturePlot(seu, features = c(\"Distal.tubule.cells_enriched\", \n",
    "                              \"Intercalated.cells_enriched\", \n",
    "                              \"Juxtaglomerular.cells_enriched\", \n",
    "                              \"Podocytes_enriched\", \n",
    "                              \"Mesangial.cells_enriched\", \n",
    "                              \"Proximal.tubule.cells_enriched\", \n",
    "                              \"Loop.of.Henle.cells_enriched\"), \n",
    "            cols = c(\"blue\", \"red\"))\n",
    "\n",
    "##Get cell type from reference dataset\n",
    "\n",
    "reference_seu <- readRDS(\"./data/Seurat Objects/reference_seu.rds\")\n",
    "\n",
    "anchors <- FindTransferAnchors(reference = reference_seu, query = seu,\n",
    "                               dims = 1:15, reference.reduction = \"pca\")\n",
    "predictions <- TransferData(anchorset = anchors, refdata = reference_seu$type,\n",
    "                            dims = 1:15)\n",
    "seu <- AddMetaData(seu, metadata = predictions)\n",
    "\n",
    "DimPlot(seu, reduction = \"umap\", label = T, group.by=\"predicted.id\", label.size = 7)\n",
    "\n",
    "##Diagnostic plots\n",
    "DimPlot(seu, label = T, shuffle = T, repel = T) + NoLegend()\n",
    "# split by sample\n",
    "DimPlot(seu, label = T, shuffle = T, repel = T, split.by = \"sample\") + NoLegend()\n",
    "# group by sample\n",
    "DimPlot(seu, label = T, shuffle = T, repel = T, group.by = \"sample\") + NoLegend()\n",
    "# QC plots\n",
    "p1 <- FeaturePlot(seu, features = \"nCount_RNA\")\n",
    "p2 <- FeaturePlot(seu, features = \"nFeature_RNA\")\n",
    "p3 <- FeaturePlot(seu, features = \"percent_mito\")\n",
    "p4 <- FeaturePlot(seu, features = \"scDblFinder.score\")\n",
    "cowplot::plot_grid(p1,p2,p3,p4, ncol = 2)\n",
    "#Barplot of cell types,\n",
    "seu$celltype <- Idents(seu)# Before that step, set identities to the cell type. \n",
    "                           # Alternatively, use the metadata column where the cell types are described.\n",
    "df <- as.data.frame(table(seu$celltype))\n",
    "\n",
    "ggplot(df, aes(Var1, Freq, fill = Var2)) + \n",
    "  geom_bar(stat = \"identity\")  + \n",
    "  theme_minimal() + \n",
    "  theme(axis.text.x = element_text(angle = 90, size = 15)) +\n",
    "  theme(axis.text.y = element_text(size = 15)) +\n",
    "  theme(axis.title = element_text(size = 15)) +\n",
    "  theme(legend.text = element_text(size = 15)) +\n",
    "  theme(title = element_text(size = 15)) +\n",
    "  theme(legend.title=element_blank()) +\n",
    "  xlab(\"Celltype\") + \n",
    "  ylab(\"Count\") +\n",
    "  ggtitle(\"Celltype distribution\") +\n",
    "  scale_fill_viridis_d()\n",
    "\n",
    "##Step 8: Differential expression\n",
    "#FindMarkers can be used for differential expression analysis. \n",
    "#There are many available methods, but for a first approach, the default can be used.\n",
    "#Specific cell types can be subset and DE analysis conducted within them.\n",
    "\n",
    "# Define function for DE within subset (in this example, mut vs all wt. The gtype metadata column contains this info.)\n",
    "FindMarkers_in_subset <- function(x, y){\n",
    "  tmp_subset <- subset(x, subset = gtype == \"wt\" | (gtype == \"mut\" & age == y))\n",
    "  \n",
    "  Idents(tmp_subset) <- tmp_subset$gtype\n",
    "  tmp_DE_table <- FindMarkers(tmp_subset, \n",
    "                              ident.1 = \"mut\",\n",
    "                              ident.2 = \"wt\",\n",
    "                              verbose = T,\n",
    "                              #min.cells.group = 50)\n",
    "                              min.cells.group = 10,\n",
    "                              min.pct = 0.1, \n",
    "                              logfc.threshold = -Inf,\n",
    "                              min.cells.feature = 1)\n",
    "  tmp_DE_table$gene <- rownames(tmp_DE_table)\n",
    "  tmp_DE_table$age <- y\n",
    "  return(tmp_DE_table)\n",
    "}\n",
    "\n",
    "### Endothelial cell markers in different ages (example)\n",
    "endo_subset <- subset(seu, idents = c(\"Endothelial\"))\n",
    "Idents(endo_subset) <- endo_subset$celltype_gtype\n",
    "\n",
    "endo_DE <- data.frame()\n",
    "\n",
    "for(a in c(4,6,8,12)){\n",
    "  tmp_DE <- FindMarkers_in_subset(endo_subset, a)\n",
    "  endo_DE <- rbind.data.frame(endo_DE, tmp_DE)\n",
    "}\n",
    "\n",
    "### Volcano Plots for FindMarkers\n",
    "tmp_obj <- endo_DE\n",
    "celltype <- \"Endothelial\"\n",
    "age = 12\n",
    "\n",
    "data <- tmp_obj[tmp_obj$age == age,] \n",
    "# Label DE genes\n",
    "data$DE <- \"not sig\"\n",
    "data$DE[data$avg_log2FC > 0.6 & data$p_val_adj < 0.01] <- \"up\" \n",
    "data$DE[data$avg_log2FC < -0.6 & data$p_val_adj < 0.01 ] <- \"down\"\n",
    "data$DElabel <- NA\n",
    "\n",
    "# Get top 10 regulated genes for labeling\n",
    "top10 <- head(arrange(data, p_val_adj), 10)\n",
    "\n",
    "# Take the highest p value of the top10 to set the threshold for labeling\n",
    "thr <- top10$p_val_adj[10]\n",
    "\n",
    "# Change labels\n",
    "data$DElabel[data$p_val_adj <= thr & !is.na(data$p_val_adj) & abs(data$avg_log2FC) > 0.6] <- data$gene[data$p_val_adj <= thr & !is.na(data$p_val_adj) & abs(data$avg_log2FC) > 0.6]\n",
    "\n",
    "# Volcano plot\n",
    "ggplot(data, aes(x = avg_log2FC, y = -log(p_val_adj), col = DE, label = DElabel)) +\n",
    "  geom_point(size = 4) +\n",
    "  theme_classic() + \n",
    "  theme(legend.position = \"none\") +\n",
    "  geom_label_repel(colour = \"black\", max.overlaps = 10, size = 5) +\n",
    "  scale_color_manual(values = c(\"blue4\", \"grey\", \"red\")) +\n",
    "  ggtitle(paste(celltype, age, \"weeks\", sep = \" \")) +\n",
    "  theme(text = element_text(size = 20), plot.title = element_text(hjust = 0.5), legend.position = \"right\")\n",
    "\n",
    "##Pseudobulk differential expression.\n",
    "#Especially useful if there are batch effects that should be modelled.\n",
    "\n",
    "DefaultAssay(seu) # The default assay should be RNA. scTransform residuals violate the assumptions of the methods\n",
    "                  # and will return many false positives.\n",
    "\n",
    "## Define Function\n",
    "# Define pseudobulk function that returns the DE analysis tables\n",
    "pseudobulk <- function(seu, weeks, celltype){ #Age and cell type are the factors of interest here, but they can\n",
    "                                              #be whatever is appropriate for the dataset.\n",
    "  seu_DE <- seu\n",
    "  # Subset seurat for condition of interest\n",
    "  seu_DE <- seu_DE %>%\n",
    "    subset(subset = gtype == \"wt\" | age == weeks) \n",
    "  \n",
    "  # Get counts for each celltype in each sample\n",
    "  counts_list <- AggregateExpression(seu_DE, \n",
    "                                     group.by = c(\"celltype\", \"sample\"),\n",
    "                                     assays = \"RNA\",\n",
    "                                     slot = \"counts\",\n",
    "                                     return.seurat = F)\n",
    "  \n",
    "  counts <- counts_list$RNA\n",
    "  # Transpose and create data frame\n",
    "  counts.t <- as.data.frame(t(counts))\n",
    "  # Create a list for each celltype \n",
    "  splitRows <- gsub(\"_.*\", \"\", rownames(counts.t))\n",
    "  # Split dataframe\n",
    "  counts.split <- split.data.frame(counts.t, \n",
    "                                   f = factor(splitRows))\n",
    "  counts.split.mod <- lapply(counts.split, function(x){\n",
    "    rownames(x) <- gsub(\".*_(.*)\", \"\\\\1\", rownames(x))\n",
    "    t(x)\n",
    "  })\n",
    "  \n",
    "#Pseudobulk DE. DESeq2 is used here, but edgeR or limma/voom are also alternatives.\n",
    "  \n",
    "  # Get count matrix of cell type\n",
    "  analysis.table <- NULL\n",
    "  counts_DE <- counts.split.mod[[celltype]]\n",
    "  \n",
    "  #Generate sample level metadata\n",
    "  colData <- data.frame(samples = colnames(counts_DE))\n",
    "  colData <- colData %>%\n",
    "    mutate(genotype = ifelse(grepl(\"mut\", samples), \"mut\", \"wt\")) %>%\n",
    "    column_to_rownames(var = \"samples\")\n",
    "  \n",
    "  ´#Create DESeq2 object\n",
    "  dds <- DESeqDataSetFromMatrix(countData = counts_DE, \n",
    "                                colData = colData,\n",
    "                                design = ~ genotype)\n",
    "  \n",
    "  #Filter: keep genes that have at least 10 reads in total\n",
    "  keep <- rowSums(counts(dds)) >= 10\n",
    "  dds <- dds[keep,]\n",
    "  \n",
    "  #Run DESEq\n",
    "  dds <- DESeq(dds)\n",
    "  \n",
    "  # Check coefficients for the comparison\n",
    "  resultsNames(dds)\n",
    "  \n",
    "  # Create results object\n",
    "  res <- results(dds, name = \"genotype_mut_vs_wt\")\n",
    "  res\n",
    "  \n",
    "  analysis.table <- as.data.frame(dplyr::arrange(as.data.frame(res), padj))\n",
    "  analysis.table$gene <- rownames(analysis.table)\n",
    "  analysis.table$age <- weeks\n",
    "  analysis.table$celltype <- celltype\n",
    "  return(analysis.table)\n",
    "}\n",
    "\n",
    "\n",
    "# Define ages\n",
    "age_of_interest <- c(4,6,8,12)\n",
    "# Define cell types\n",
    "celltypes_of_interest <- c(\"Endothelial\", \"Mesangial\", \"Immune\", \"Podocytes\")\n",
    "# create empty analysis table list\n",
    "analysis_table_list <- list()\n",
    "\n",
    "for(c in celltypes_of_interest){\n",
    "  print(c)\n",
    "  tmp_df <- data.frame()\n",
    "  for(a in age_of_interest){\n",
    "    print(paste(c, a, \"weeks\", sep = \" \"))\n",
    "    tmp_analysis_table <- pseudobulk(seu, a, c)\n",
    "    tmp_df <- rbind(tmp_df, tmp_analysis_table)\n",
    "  }\n",
    "  analysis_table_list[[c]] <- tmp_df\n",
    "  gc()\n",
    "}\n",
    "\n",
    "#Similarly to FindMarkers, volcano plots can be created to check differentially expressed genes.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
